[opaque_obj] Add nested value-type opaque obj support

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169845
* #169846
* #167768

In the case of DTensorSpec, it contains multiple opaque objects like DeviceMesh and Placements. So when creating the repr, we need to get the list of nested types too to be added to the FX graph or inductor codegen.

The graph with nested value types would look something like:
```
def forward(self, arg0_1):
    process_nested_config = torch.ops._TestOpaqueObject.process_nested_config.default(arg0_1, NestedValueSize(size=SizeStore(size=3), config=ValueConfig(mode='square')));  arg0_1 = None
    return (process_nested_config,)
```


cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela

angelayi committed:
[opaque_obj] Add nested value-type opaque obj support

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Add nested value-type opaque obj support"




cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Add nested value-type opaque obj support"


In the case of DTensorSpec, it contains multiple opaque objects like DeviceMesh and Placements. So when creating the repr, we need to get the list of nested types too to be added to the FX graph or inductor codegen.

The graph with nested value types would look something like:
```
def forward(self, arg0_1):
    process_nested_config = torch.ops._TestOpaqueObject.process_nested_config.default(arg0_1, NestedValueSize(size=SizeStore(size=3), config=ValueConfig(mode='square')));  arg0_1 = None
    return (process_nested_config,)
```


cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Add nested value-type opaque obj support"


In the case of DTensorSpec, it contains multiple opaque objects like DeviceMesh and Placements. So when creating the repr, we need to get the list of nested types too to be added to the FX graph or inductor codegen.

The graph with nested value types would look something like:
```
def forward(self, arg0_1):
    process_nested_config = torch.ops._TestOpaqueObject.process_nested_config.default(arg0_1, NestedValueSize(size=SizeStore(size=3), config=ValueConfig(mode='square')));  arg0_1 = None
    return (process_nested_config,)
```


cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Add nested value-type opaque obj support"


In the case of DTensorSpec, it contains multiple opaque objects like DeviceMesh and Placements. So when creating the repr, we need to get the list of nested types too to be added to the FX graph or inductor codegen.

The graph with nested value types would look something like:
```
def forward(self, arg0_1):
    process_nested_config = torch.ops._TestOpaqueObject.process_nested_config.default(arg0_1, NestedValueSize(size=SizeStore(size=3), config=ValueConfig(mode='square')));  arg0_1 = None
    return (process_nested_config,)
```


cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Add nested value-type opaque obj support"


In the case of DTensorSpec, it contains multiple opaque objects like DeviceMesh and Placements. So when creating the repr, we need to get the list of nested types too to be added to the FX graph or inductor codegen.

The graph with nested value types would look something like:
```
def forward(self, arg0_1):
    process_nested_config = torch.ops._TestOpaqueObject.process_nested_config.default(arg0_1, NestedValueSize(size=SizeStore(size=3), config=ValueConfig(mode='square')));  arg0_1 = None
    return (process_nested_config,)
```


cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

angelayi committed:
Update on "[opaque_obj] Add nested value-type opaque obj support"


In the case of DTensorSpec, it contains multiple opaque objects like DeviceMesh and Placements. So when creating the repr, we need to get the list of nested types too to be added to the FX graph or inductor codegen.

The graph with nested value types would look something like:
```
def forward(self, arg0_1):
    process_nested_config = torch.ops._TestOpaqueObject.process_nested_config.default(arg0_1, NestedValueSize(size=SizeStore(size=3), config=ValueConfig(mode='square')));  arg0_1 = None
    return (process_nested_config,)
```


cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]

zou3519 reviewed (approved):
lgtm, though I might like `__fx_repr__` more than `__fx_eval__` (feel free to keep bikeshedding tho)

angelayi committed:
Update on "[opaque_obj] Add nested value-type opaque obj support"


In the case of DTensorSpec, it contains multiple opaque objects like DeviceMesh and Placements. So when creating the repr, we need to get the list of nested types too to be added to the FX graph or inductor codegen.

The graph with nested value types would look something like:
```
def forward(self, arg0_1):
    process_nested_config = torch.ops._TestOpaqueObject.process_nested_config.default(arg0_1, NestedValueSize(size=SizeStore(size=3), config=ValueConfig(mode='square')));  arg0_1 = None
    return (process_nested_config,)
```


cc ezyang EikanWang jgong5 wenzhe-nrv voznesenskym penguinwu Guobing-Chen XiaobingSuper zhuhaozhe blzheng jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo

[ghstack-poisoned]
